libkbfs/mdserver_disk.go

```diff
@@ -100,70 +76,55 @@ func NewMDServerLocal(config Config, handleDbfile string, mdDbfile string,
+	handleDb, err := leveldb.Open(handleStorage, leveldbOptions)
+	if err != nil {
+		return nil, err
 	}
-	_, user, err := md.config.KBPKI().GetCurrentUserInfo(ctx)
+	mdDb, err := leveldb.Open(mdStorage, leveldbOptions)
 	if err != nil {
-		return false, err
+		return nil, err
 	}
-	h, err := rmds.MD.MakeBareTlfHandle()
+	branchDb, err := leveldb.Open(branchStorage, leveldbOptions)
 	if err != nil {
-		return false, err
+		return nil, err
 	}
-	isWriter := h.IsWriter(user)
-	isReader := h.IsReader(user)
-	if checkWrite {
-		// if this is a reader, are they acting within their restrictions?
-		if !isWriter && isReader && newMd != nil {
-			return newMd.MD.IsValidRekeyRequest(md.config, &rmds.MD, user)
-		}
-		return isWriter, nil
+	locksDb, err := leveldb.Open(lockStorage, leveldbOptions)
+	if err != nil {
+		return nil, err
 	}
-	return isWriter || isReader, nil
-}
-
-// Helper to aid in enforcement that only specified public keys can access TLF metdata.
-func (md *MDServerLocal) isReader(ctx context.Context, id TlfID) (bool, error) {
-	return md.checkPerms(ctx, id, false, nil)
+	log := config.MakeLogger("")
+	mdserv := &MDServerDisk{config, handleDb, mdDb, branchDb, log,
+		&sync.Mutex{}, locksDb, newMDServerLocalUpdateManager(),
+		&sync.RWMutex{}, new(bool), shutdownFunc}
+	return mdserv, nil
 }
 
-// Helper to aid in enforcement that only specified public keys can access TLF metdata.
-func (md *MDServerLocal) isWriter(ctx context.Context, id TlfID) (bool, error) {
-	return md.checkPerms(ctx, id, true, nil)
+// NewMDServerDir constructs a new MDServerDisk that stores its data
+// in the given directory.
+func NewMDServerDir(config Config, dirPath string) (*MDServerDisk, error) {
+	return newMDServerDisk(config, dirPath, nil)
 }
 
-// Helper to aid in enforcement that only specified public keys can access TLF metdata.
-func (md *MDServerLocal) isWriterOrValidRekey(ctx context.Context, id TlfID, newMd *RootMetadataSigned) (
-	bool, error) {
-	return md.checkPerms(ctx, id, true, newMd)
+// NewMDServerTempDir constructs a new MDServerDisk that stores its
+// data in a temp directory which is cleaned up on shutdown.
+func NewMDServerTempDir(config Config) (*MDServerDisk, error) {
+	tempdir, err := ioutil.TempDir(os.TempDir(), "kbfs_mdserver_tmp")
+	if err != nil {
+		return nil, err
+	}
+	return newMDServerDisk(config, tempdir, func(log logger.Logger) {
+		err := os.RemoveAll(tempdir)
+		if err != nil {
+			log.Warning("error removing %s: %s", tempdir, err)
+		}
+	})
 }
 
-// GetForHandle implements the MDServer interface for MDServerLocal.
-func (md *MDServerLocal) GetForHandle(ctx context.Context, handle BareTlfHandle,
+// GetForHandle implements the MDServer interface for MDServerDisk.
+func (md *MDServerDisk) GetForHandle(ctx context.Context, handle BareTlfHandle,
 	mStatus MergeStatus) (TlfID, *RootMetadataSigned, error) {
 	id := NullTlfID
-	md.shutdownLock.RLock()
-	defer md.shutdownLock.RUnlock()
-	if *md.shutdown {
+	if md.isShutdown() {
```

Comment:
This (and other functions below) no longer holds the shutdown lock throughout the call, so a shutdown could sneak in right after this check and destroy the storage, right?